File: lexer/select.go
package lexer

import (
    "strings"
    "fmt"
)

func TokeniseSelect(query string) []string {

	tokens := []string{}
	queryString := strings.Split(query, " ")

    lastAdded := ""
	for i, word := range queryString {

		if stringInSlice(word, token["keyword"]) {
			tokens = append(tokens, []string{"keyword", word}...)
        } else if i > 0 && (lastAdded == "SELECT" || lastAdded == "FROM") {
			tokens = append(tokens, []string{"table_reference", word}...)
		} else {
			//TODO once we're happy with the structure as a whole, turn this to a panic
			tokens = append(tokens, []string{"???", word}...)
		}
        lastAdded = word
	}

    fmt.Println("")
    return tokens
}

File: src/lexer/select.ts
import { Query } from "../reader/reader";
import { ILexer } from "./interface";
import { cleanUnquotedIdentifier } from "./lexer";
import { TOKENS, Keyword} from "./tokens";

class Select implements ILexer {
  public tokenise(query: Query): Query {
    let lastToken = "";
    query.lines.forEach(line => {
      line.content.split(" ").forEach(word => {
        let item = word.toLowerCase();

        if (TOKENS.keyword.includes(item)) {
          line.tokens.push(["keyword", item]);
        } else if (lastToken === Keyword.Select || lastToken === Keyword.From) {
          item = cleanUnquotedIdentifier(item);

          if (item.length > 0) {
            line.tokens.push([
              "table_reference",
              cleanUnquotedIdentifier(item)
            ]);
          }
        } else {
          line.tokens.push(["???", item]);
        }
        lastToken = item;
      });
    });

    return query;
  }
}

export { Select };
